codec-http2/src/test/java/io/netty/handler/codec/http2/Http2CodecTest.java

```diff
@@ -0,0 +1,159 @@
+        serverChannel = sb.bind(serverAddress).sync().channel();
+
+        Bootstrap cb = new Bootstrap()
+                .channel(LocalChannel.class)
+                .group(group)
+                .handler(new Http2Codec(false, new TestChannelInitializer()));
+        clientChannel = cb.connect(serverAddress).sync().channel();
+    }
+
+    @AfterClass
+    public static void shutdown() {
+        group.shutdownGracefully();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        clientChannel.close().sync();
+        serverChannel.close().sync();
+    }
+
+    @Test
+    public void multipleOutboundStreams() throws InterruptedException{
+        Http2StreamChannelBootstrap b = new Http2StreamChannelBootstrap();
+        b.parentChannel(clientChannel).handler(new TestChannelInitializer());
+
+        Channel childChannel1 = b.connect().sync().channel();
+        assertTrue(childChannel1.isActive());
+        assertFalse(((AbstractHttp2StreamChannel) childChannel1).hasStreamId());
+        Channel childChannel2 = b.connect().channel();
+        assertTrue(childChannel2.isActive());
+        assertFalse(((AbstractHttp2StreamChannel) childChannel2).hasStreamId());
+
+        Http2Headers headers1 = new DefaultHttp2Headers();
+        Http2Headers headers2 = new DefaultHttp2Headers();
+        // Test that streams can be made active (headers sent) in different order than the corresponding channels
+        // have been created.
+        childChannel2.writeAndFlush(new DefaultHttp2HeadersFrame(headers2));
+        childChannel1.writeAndFlush(new DefaultHttp2HeadersFrame(headers1));
+
+        Http2HeadersFrame headersFrame2 = serverLastInboundHandler.blockingReadInbound();
+        assertNotNull(headersFrame2);
+        assertEquals(3, headersFrame2.streamId());
+
+        Http2HeadersFrame headersFrame1 = serverLastInboundHandler.blockingReadInbound();
+        assertNotNull(headersFrame1);
+        assertEquals(5, headersFrame1.streamId());
+
+        assertTrue(((AbstractHttp2StreamChannel) childChannel2).hasStreamId());
+        assertTrue(((AbstractHttp2StreamChannel) childChannel2).hasStreamId());
+
+        childChannel1.close();
+        childChannel2.close();
+    }
+
+    @Test
+    public void createOutboundStream() throws InterruptedException {
+        Http2StreamChannelBootstrap b = new Http2StreamChannelBootstrap();
+        Channel childChannel = b.parentChannel(clientChannel).handler(new TestChannelInitializer())
+                                       .connect().sync().channel();
+        assertTrue(childChannel.isRegistered());
+        assertTrue(childChannel.isActive());
+
+        Http2Headers headers = new DefaultHttp2Headers();
+        childChannel.write(new DefaultHttp2HeadersFrame(headers));
+        ByteBuf data = Unpooled.buffer(100).writeZero(100);
+        childChannel.writeAndFlush(ReferenceCountUtil.releaseLater(new DefaultHttp2DataFrame(data, true).retain()));
```

Comment:
I think I may have confused things when I was describing typical behavior of EmbeddedChannel above. Typically a write is done with no `retain()`, then the object is released when it is read out. This simulates a "regular write" and a release by the transport (see `HttpContentEncoderTest`)